### PR TITLE
r/elasticsearch_domain - add support for elasticsearch warm storage

### LINKED ISSUE #12606: Support for Ultrawarm with aws_elasticsearch_domain
Community Note
Please vote on this issue by adding a 👍 reaction to the original issue to help the community and maintainers prioritize this request
Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
If you are interested in working on this issue or have submitted a pull request, please leave a comment
Description
AWS ES Team Launched UW https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/ultrawarm.html
These option to be added to https://www.terraform.io/docs/providers/aws/r/elasticsearch_domain.html 
    "WarmEnabled": true,
    "WarmCount": 6,
    "WarmType": "ultrawarm1.medium.elasticsearch"

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12606

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_elasticsearch_domain: add support for elasticsearch warm storage
resource_aws_elasticsearch_domain: add plan time validation for `cloudwatch_log_group_arn`, `role_arn`, and `ebs_options.volume_type`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_warm'
--- PASS: TestAccAWSElasticSearchDomain_warm (1737.90s)

```

### CODE REVIEW

**Comment by leoskyrocker** [comment]
Who can review this?

File: aws/resource_aws_elasticsearch_domain_test.go [hunk 1]
```diff
@@ -1065,6 +1101,36 @@ resource "aws_elasticsearch_domain" "test" {
...
 
+func testAccESDomainConfigWarm(rName string, enabled bool, count int) string {
+	return fmt.Sprintf(`
+resource "aws_elasticsearch_domain" "test" {
+  domain_name           = %[1]q
+  elasticsearch_version = "6.8"
+
+  cluster_config {
+    zone_awareness_enabled   = true
+    instance_type            = "r5.large.elasticsearch"
```
Comment (tomelliff) [inline aws/resource_aws_elasticsearch_domain_test.go:1112]: This and the master type seem overly large for running an acceptance test. Any reason not to use the smallest available cluster to test the ultrawarm stuff?

From the [docs](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/ultrawarm.html#ultrawarm-pp) it looks like you can't have t2 nodes as data nodes and must have some form of master nodes to be able to use ultrawarm so I think the [cheapest](https://aws.amazon.com/elasticsearch-service/pricing/) cluster size here is a single AZ c5.large data node and a t2.micro master node.

File: aws/resource_aws_elasticsearch_domain.go [hunk 2]
```diff
@@ -212,7 +212,7 @@ func resourceAwsElasticSearchDomain() *schema.Resource {
 						"instance_type": {
 							Type:     schema.TypeString,
 							Optional: true,
-							Default:  "m3.medium.elasticsearch",
+							Default:  elasticsearch.ESPartitionInstanceTypeM3MediumElasticsearch,
```
Comment (tomelliff) [inline aws/resource_aws_elasticsearch_domain.go:215]: This is a nice catch but I am slightly confused why this is defaulted anyway and why to this specific instance type when I can't see any mention of it in the AWS docs at all.

Unfortunately removing this default would be a breaking change :crying_cat_face:

File: aws/resource_aws_elasticsearch_domain_test.go (hunk 1 above)
Comment (DrFaust92) [reply aws/resource_aws_elasticsearch_domain_test.go:1112]: will test these

Comment (ace03uec) [reply aws/resource_aws_elasticsearch_domain_test.go:1112]: AWS Console doesn't allow creation of ultrawarm nodes with t2 master nodes, as well.

Comment (DrFaust92) [reply aws/resource_aws_elasticsearch_domain_test.go]: changed, and as @ace03uec said. master node cannot be t2 for ultrawarm

**Comment by DrFaust92** [comment]
Hey @gdavison, thanks for the comments, I'll add more info to docs and add checks for the cases you mentioned.

File: website/docs/r/elasticsearch_domain.html.markdown [hunk 3]
```diff
@@ -256,6 +256,9 @@ The following arguments are supported:
 * `dedicated_master_count` - (Optional) Number of dedicated master nodes in the cluster
 * `zone_awareness_config` - (Optional) Configuration block containing zone awareness settings. Documented below.
 * `zone_awareness_enabled` - (Optional) Indicates whether zone awareness is enabled, set to `true` for multi-az deployment. To enable awareness with three Availability Zones, the `availability_zone_count` within the `zone_awareness_config` must be set to `3`.
+* `warm_enabled` - (Optional) Indicates whether to enable warm storage.
+* `warm_count` - (Optional) The number of warm nodes in the cluster.
```
[the remark this answers is not in the document]
Comment (DrFaust92) [reply website/docs/r/elasticsearch_domain.html.markdown]: done

File: aws/resource_aws_elasticsearch_domain.go [hunk 4]
```diff
@@ -239,6 +239,24 @@ func resourceAwsElasticSearchDomain() *schema.Resource {
...
 							Optional: true,
 						},
+						"warm_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							ForceNew: true,
+						},
+						"warm_count": {
+							Type:         schema.TypeInt,
+							Optional:     true,
```
[the remark this answers is not in the document]
Comment (DrFaust92) [reply aws/resource_aws_elasticsearch_domain.go:249]: validation exception is thrown

**Comment by DrFaust92** [comment]
Hey @gdavison, ive addressed the comments and rerun test
`--- PASS: TestAccAWSElasticSearchDomain_warm (3273.30s)`

File: aws/resource_aws_elasticsearch_domain.go [hunk 5]
```diff
@@ -239,6 +239,25 @@ func resourceAwsElasticSearchDomain() *schema.Resource {
 							Type:     schema.TypeBool,
 							Optional: true,
 						},
+						"warm_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							ForceNew: true,
```
Comment (amulyas) [inline aws/resource_aws_elasticsearch_domain.go:245]: Why ForceNew is set to true?

**Pull request merged**